selenium_test.py

```diff
@@ -0,0 +1,222 @@
+
+        if len(enabledButton) > 0:
+            enabledButton[0].click()
+            collectResults(driver, keyword)
+            hasNextPage = True
+        elif len(disabledButton) > 0:
+            collectResults(driver, keyword)
+            break
+
+    # waits for the new rollover download
+    markedWait = WebDriverWait(driver, 10).until(
+        EC.presence_of_element_located((By.CLASS_NAME, "markListSpan"))
+    )
+
+    # clicks marked items button
+    markedItems = driver.find_element_by_class_name("markListSpan")
+    markedItems.click()
+
+    # clicks marked items button
+    download = driver.find_element_by_id("download")
+    download.click()
+
+    currentWindow = driver.current_window_handle
+    downloadPage = driver.window_handles[1]
+    driver.switch_to.window(downloadPage)
+
+    # waits for the new rollover download
+    resultsWait = WebDriverWait(driver, 10).until(
+        EC.presence_of_element_located((By.CLASS_NAME, "rolloverdownload"))
+    )
+
+    # clicks marked items button
+    rolloverdownload = driver.find_element_by_class_name("rolloverdownload")
+    rolloverdownload.click()
+
+    driver.switch_to.window(currentWindow)
+
+    df = pd.DataFrame(list(zip(lists.title_list, lists.description_list, lists.category_list, lists.keyword_list, lists.link_list)),
+                      columns=['Title', 'Description', 'Category', 'Keyword', "link"])
+
+    df.to_csv('selenium-output.csv', index=True)
+
+
+
+
+
+
+
+
+
+
+def collectResults(driver, keyword):
+    # waits for the presence of the Ul item  on new results page
+    time.sleep(1)
+    resultsWait = WebDriverWait(driver, 30).until(
+        EC.presence_of_element_located((By.CLASS_NAME, "resultsListBox"))
+    )
+
+    # grabs list item search results
+    resultItems = driver.find_elements_by_class_name("resultsListBox")
+
+
+    for result in resultItems:
+        title = result.find_element_by_class_name("articleTitle").text
+        description = result.find_element_by_class_name("description").text
+        category = result.find_element_by_class_name("articleType").text
+        link = result.find_element_by_class_name("articleTitle").find_element_by_css_selector('a').get_attribute('href')
+
+
+
+
+
+
+
+
+        lists.link_list.append(link)
+        lists.title_list.append(title)
+        lists.description_list.append(description)
+        lists.category_list.append(category)
+        lists.keyword_list.append(keyword)
+
+
+        markItem = result.find_element_by_class_name("markItem")
+        markItem.click()
+
+    return lists
+
+
+
+
+
+
+def groupResults():
+    df = pd.read_csv('selenium-output.csv', usecols=['Title','Description','Category','Keyword',"link"])
+    #print(df)
+    #df.columns = df.columns.str.strip()
+    df['length'] = df['Keyword'].str.len()
+    out = df.astype(str).groupby(['Title']).agg(', '.join)
+    out.sort_values('length', ascending=False, inplace=True)
```

Comment:
You also want to do the same sorting and ranking of articles here, just like we did with Scrapy.

Reply:
1. I don't think `.agg(', '.join)` does what you want - wouldn't it also join the description and other fields that you don't want joined?

2. I don't think this ranks the links by decreasing number of matching keywords, does it? The length is an approximation, but it would fail if I have a document that matches 'aaa' and 'bbb', and another document that matches 'cccccccccccc'.

Reply:
Yes it would join the description and other fields if it were to join articles that are not the same. However, since I am joining by grouping the titles, it will only group those together that are the same article but under a different keyword search. So, the only thing that ends up being combined is the keyword

Reply:
I'm not convinced:

```python
>>> import pandas as pd

>>> df = pd.DataFrame([['hello', 'foo', '1'], ['hello', 'foo', '2']], columns=['url', 'description', 'keyword'])

>>> df

     url description keyword
0  hello         foo       1
1  hello         foo       2

>>> df.groupby(['url']).agg(', '.join)

      description keyword
url
hello    foo, foo    1, 2
>>>
```